File: clients/csharp/Runner.cs
using System;
using System.Collections.Generic;
using System.Diagnostics.CodeAnalysis;
using System.IO;
using System.Net.Sockets;
using System.Text;

namespace AiCup22
{
    public class Runner
    {
        private BinaryReader reader;
        private BinaryWriter writer;
        public Runner(string host, int port, string token)
        {
            var client = new TcpClient(host, port) { NoDelay = true };
            var stream = new BufferedStream(client.GetStream());
            reader = new BinaryReader(stream);
            writer = new BinaryWriter(stream);
            var tokenData = System.Text.Encoding.UTF8.GetBytes(token);
            writer.Write(tokenData.Length);
            writer.Write(tokenData);
            writer.Write((int) 1);
            writer.Write((int) 0);
            writer.Write((int) 1);
            writer.Flush();
        }
        public void Run()
        {
            MyStrategy myStrategy = null;
            var debugInterface = new DebugInterface(reader, writer);
            var running = true;
            while (running)
            {
                switch (AiCup22.Codegame.ServerMessage.ReadFrom(reader))
                {
                    case AiCup22.Codegame.ServerMessage.UpdateConstants message:
                        myStrategy = new MyStrategy(message.Constants);
                        break;
                    case AiCup22.Codegame.ServerMessage.GetOrder message:
                        new AiCup22.Codegame.ClientMessage.OrderMessage(myStrategy.GetOrder(message.PlayerView, message.DebugAvailable ? debugInterface : null)).WriteTo(writer);
                        writer.Flush();
                        break;
                    case AiCup22.Codegame.ServerMessage.Finish message:
                        running = false;
                        myStrategy.Finish();
                        break;
                    case AiCup22.Codegame.ServerMessage.DebugUpdate message:
                        myStrategy.DebugUpdate(debugInterface);
                        new AiCup22.Codegame.ClientMessage.DebugUpdateDone().WriteTo(writer);
                        writer.Flush();
                        break;
                    default:
                        throw new Exception("Unexpected server message");
                }
            }
        }
        public static void Main(string[] args)
        {
            string host = args.Length < 1 ? "127.0.0.1" : args[0];
            int port = args.Length < 2 ? 31001 : int.Parse(args[1]);
            string token = args.Length < 3 ? "0000000000000000" : args[2];
            new Runner(host, port, token).Run();
        }
    }
}

File: clients/fsharp/Runner.fs
namespace AiCup22

open System
open System.IO
open System.Net.Sockets

module Runner =
    type T(host, port, token: string) =
        let client = new TcpClient(host, port)
        let stream = new BufferedStream(client.GetStream())
        let reader = new BinaryReader(stream)
        let writer = new BinaryWriter(stream)
        let tokenData = System.Text.Encoding.UTF8.GetBytes token

        do
            client.NoDelay <- true
            writer.Write tokenData.Length
            writer.Write tokenData
            writer.Write 1
            writer.Write 0
            writer.Write 1
            writer.Flush()

        member this.run =
            let mutable myStrategy: option<MyStrategy> = None
            let debugInterface = new DebugInterface(reader, writer)

            let rec loop () =
                match Codegame.ServerMessage.readFrom reader with
                | Codegame.ServerMessage.UpdateConstants message ->
                    myStrategy <- Some(new MyStrategy(message.Constants))
                    loop()
                | Codegame.ServerMessage.GetOrder message ->
                    (Codegame.ClientMessage.OrderMessage { Order = myStrategy.Value.getOrder(message.PlayerView, (if message.DebugAvailable then Some debugInterface else None)) }).writeTo writer
                    writer.Flush()
                    loop ()
                | Codegame.ServerMessage.Finish message ->
                    myStrategy.Value.finish()
                | Codegame.ServerMessage.DebugUpdate message ->
                    myStrategy.Value.debugUpdate (debugInterface)
                    (new Codegame.ClientMessageDebugUpdateDone()).writeTo writer
                    writer.Flush()
                    loop ()

            loop ()

    [<EntryPoint>]
    let main argv =
        let host =
            match argv with
            | x when x.Length >= 1 -> x.[0]
            | _ -> "127.0.0.1"

        let port =
            match argv with
            | x when x.Length >= 2 -> x.[1] |> Int32.Parse
            | _ -> 31001

        let token =
            match argv with
            | x when x.Length >= 3 -> x.[2]
            | _ -> "0000000000000000"

        (new T(host, port, token)).run

        0 // return an integer exit code